### ISSUE TITLE
[nightly regression] E0639 + adding #[non_exhaustive] to io::ErrorKind is a breaking change

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#53549]
In stable and beta, you can write `ErrorKind::Other {}`, but in nightly this triggers E0639: cannot create non-exhaustive variant using struct expression. (As an aside, that error message should be rephrased and examples added to `--explain`.)

This is either a bug with `#[non_exhaustive]`, or it means that the attribute can't be applied to public enums without creating a breaking change.

**Comment by durka** [comment]
I noticed this due to Travis tests of [this code here](https://github.com/durka/macrolisp/blob/e3719827c4ef22bfb4c51ff8a47dc2486c08e71e/examples/tokio_echo.rs#L41). This is obviously(?) not a serious crate, but illustrates that this restriction is problematic for macros. That line expands to

```rust
    Err(_) => Err(io::Error::new(io::ErrorKind::Other{}, "invalid UTF-8"))
```

**Comment by Havvy** [comment]
This is a bug. Adding `#[non_exhaustive]` to an enum should not prevent the creation of enum variants, although adding the attribute to an enum variant explicitly allows for that. Especially since `ErrorKind::Other` without the `{}` is still valid.

**Comment by arielb1** [comment]
I have a fix

**Comment by pnkfelix** [comment]
visiting for triage. P-high. (Fixed by #53721 in any case)
